candle-core/src/metal_backend/device.rs

```diff
@@ -99,48 +132,85 @@ impl MetalDevice {
-            *command_buffer_lock = command_buffer.clone();
-            *index = 0;
+        let command_buffer = command_buffer_lock.to_owned();
+
+        if let Some(command_buffer) = command_buffer {
+            Ok(command_buffer)
+        } else {
+            let new_command_buffer = self.command_queue.new_command_buffer().to_owned();
+            *command_buffer_lock = Some(new_command_buffer);
+            Ok(command_buffer_lock.to_owned().unwrap())
+        }
+    }
 
-            self.drop_unused_buffers()?;
+    /// Returns the current active command encoder, if the encoder is not present, this will allocate one
+    pub fn command_encoder(&self) -> Result<ComputeCommandEncoder> {
+        let mut command_encoder_lock =
+            self.command_encoder.try_write().map_err(MetalError::from)?;
+        let command_encoder = command_encoder_lock.to_owned();
+
+        if let Some(command_encoder) = command_encoder {
+            Ok(command_encoder)
+        } else {
+            let command_buffer = self.command_buffer()?;
+            let new_command_encoder = command_buffer.new_compute_command_encoder().to_owned();
+            *command_encoder_lock = Some(ComputeCommandEncoder::from(new_command_encoder));
+            Ok(command_encoder_lock.to_owned().unwrap())
         }
-        *index += 1;
-        Ok(command_buffer)
     }
 
-    pub fn wait_until_completed(&self) -> Result<()> {
-        let mut command_buffer = self.command_buffer.try_write().map_err(MetalError::from)?;
-        match command_buffer.status() {
-            metal::MTLCommandBufferStatus::Committed
-            | metal::MTLCommandBufferStatus::Scheduled
-            | metal::MTLCommandBufferStatus::Completed => {
-                panic!("Already committed");
-            }
-            _ => {}
+    /// Ends the current command buffer
+    pub fn close_compute_buffer(&self) -> Result<()> {
+        let mut command_buffer_lock = self.command_buffer.try_write().map_err(MetalError::from)?;
+        let command_buffer = command_buffer_lock.to_owned();
+
+        if let Some(command_buffer) = command_buffer {
+            command_buffer.commit();
+            command_buffer.wait_until_completed();
+            *command_buffer_lock = None;
         }
-        command_buffer.commit();
-        command_buffer.wait_until_completed();
-        *command_buffer = self.command_queue.new_command_buffer().to_owned();
 
         Ok(())
     }
 
-    pub fn kernels(&self) -> &Kernels {
-        &self.kernels
+    /// If the user wants to end the current command encoder, they should call this function
+    /// In the case that there is no active command encoder, this function will not return an error,
+    /// it instead is a no-op.
+    pub fn end_compute_encoding(&self) -> Result<()> {
+        let mut command_encoder_lock =
+            self.command_encoder.try_write().map_err(MetalError::from)?;
+        let command_encoder = command_encoder_lock.to_owned();
```

Comment:
Do we really need this `to_owned` here, seems like this would trigger some obj-c refcounting that may be quite wasteful whereas we only need a reference to it.